Mailchimp resubscribe

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Feature

## Description

If user's who previously subscribed to our newsletter try to resubscribe, it won't work without them explicitly opting in again. Setting their user status to 'pending' should™ trigger a confirmation email and this PR implements this.

There are several things that could be improved once we get the basics working, I'll leave some comments inline/below.

## Related Tickets & Documents

https://github.com/thepracticaldev/tech-private/issues/398

## Added tests?

- [X] yes

## Added to documentation?

- [X] no documentation needed

![complicated](https://media.giphy.com/media/cnuQwZ8IFLDZFwreWF/giphy.gif)

citizen428 commented on app/labor/mailchimp_bot.rb:
```diff
@@ -39,6 +39,24 @@ def upsert_to_newsletter
+      success = true
+    rescue Gibbon::MailChimpError => e
+      # If user was previously subscribed, set their status to "pending"
+      return resubscribe_to_newsletter if previously_subcribed?(e)
```
We could keep of whether or not the user was previously subscribed, but for now let's just inspect the error and try a resubscribe until we're sure this works as expected.

maestromac replied:
I looked through the MailChimp errors we've logged. It's half "Too many request" and half "Member in Compliance state". That potentially means that we will be sending all these users a MailChimp opt-in email without them actually toggling their preference from false to true. It might be best for us not to attempt to re-subscribe here.

citizen428 commented on app/labor/mailchimp_bot.rb:
```diff
@@ -170,4 +188,8 @@ def target_md5_email
   end
+
+  def previously_subcribed?(error)
+    error.title.match?(/Member In Compliance State/)
```
Once we confirmed resubscribing via "pending" is indeed what we want, we can revisit this check and maybe come up with something smarter.

citizen428 commented:
The biggest problem with this approach is that our view of the world may diverge from MailChimp's:

* User updates preferences on our end
* We update the DB from false to true
* The re-subscribe action fails
* Result: we consider the user subscribed while MC doesn't.

However, until we're sure that this approach is working, I did not want to change the user table to use enums instead of booleans or something like that. Let's get a basic resubscribe flow into a working state first, then we can iterate on that.

maestromac commented:
I haven't verify out my theory yet but my concern now is that we receive `Member In Compliance State` fairly frequently with the help of [my previous pr](https://github.com/thepracticaldev/dev.to/pull/6186). I didn't expect to see this many and I wonder if the logic we have to place to upsert to Mailchimp is flawed causing unnecessary updates. Perhaps there's an unaccounted callback. I will take a deeper look at development tomorrow.

citizen428 commented:
Thanks @maestromac, the whole MC setup is fairly opaque to me, so any help is appreciated. Note that the compliance state can have several reasons, not only unsubscribes:

> “XXXXX@xxxx.com is in a compliance state due to unsubscribe, bounce, or compliance review and cannot be subscribed.”

citizen428 commented:
@maestromac Having a look through the code, in `user.rb` we have the following callback:

```ruby
after_save  :subscribe_to_mailchimp_newsletter
```

The corresponding method looks like this:

```ruby
def subscribe_to_mailchimp_newsletter
  return unless email.present? && email.include?("@")
  return if saved_changes["unconfirmed_email"] && saved_changes["confirmation_sent_at"]

  Users::SubscribeToMailchimpNewsletterWorker.perform_async(id)
end
```

There appears to be no early exit in case the email address has not changed. In the following example, I just updated a `users website_url`:

```ruby
[2] dev.to(main)> user.update(website_url: "http://example.com")
   (1.1ms)  BEGIN
  BanishedUser Exists (1.5ms)  SELECT  1 AS one FROM "banished_users" WHERE "banished_users"."username" = $1 LIMIT $2  [["username", "dsfsfad"], ["LIMIT", 1]]
  User Update (1.1ms)  UPDATE "users" SET "updated_at" = $1, "website_url" = $2 WHERE "users"."id" = $3  [["updated_at", "2020-02-25 03:23:40.884120"], ["website_url", "http://example.com"], ["id", 4]]

From: /Users/michi/src/dev.to/app/models/user.rb @ line 422 User#subscribe_to_mailchimp_newsletter:

    417: def subscribe_to_mailchimp_newsletter
    418:   return unless email.present? && email.include?("@")
    419:   return if saved_changes["unconfirmed_email"] && saved_changes["confirmation_sent_at"]
    420:
    421:   binding.pry
 => 422:   Users::SubscribeToMailchimpNewsletterWorker.perform_async(id)
    423: end
```

I then inlined the call to `MailchimpBot.new(user).upsert` and put another `binding.pry` in `MailchimpBot#upsert_to_newsletter` (and removed the environment check in `upsert`). We get there just fine and toggle the subscription status in the payload:

```ruby
status: user.email_newsletter ? "subscribed" : "unsubscribed",
```

This leads me to believe that we probably need a guard clause like this in `upsert_to_newsletter`:

```ruby
@maestromac Having a look through the code, in `user.rb` we have the following callback:

```ruby
after_save  :subscribe_to_mailchimp_newsletter
```

The corresponding method looks like this:

```ruby
def subscribe_to_mailchimp_newsletter
  return unless email.present? && email.include?("@")
  return if saved_changes["unconfirmed_email"] && saved_changes["confirmation_sent_at"]

  Users::SubscribeToMailchimpNewsletterWorker.perform_async(id)
end
```

There appears to be no early exit in case the email address has not changed. In the following example, I just updated a `users website_url`:

```ruby
[2] dev.to(main)> user.update(website_url: "http://example.com")
   (1.1ms)  BEGIN
  BanishedUser Exists (1.5ms)  SELECT  1 AS one FROM "banished_users" WHERE "banished_users"."username" = $1 LIMIT $2  [["username", "dsfsfad"], ["LIMIT", 1]]
  User Update (1.1ms)  UPDATE "users" SET "updated_at" = $1, "website_url" = $2 WHERE "users"."id" = $3  [["updated_at", "2020-02-25 03:23:40.884120"], ["website_url", "http://example.com"], ["id", 4]]

From: /Users/michi/src/dev.to/app/models/user.rb @ line 422 User#subscribe_to_mailchimp_newsletter:

    417: def subscribe_to_mailchimp_newsletter
    418:   return unless email.present? && email.include?("@")
    419:   return if saved_changes["unconfirmed_email"] && saved_changes["confirmation_sent_at"]
    420:
    421:   binding.pry
 => 422:   Users::SubscribeToMailchimpNewsletterWorker.perform_async(id)
    423: end
```

I then inlined the call to `MailchimpBot.new(user).upsert` and put another `binding.pry` in `MailchimpBot#upsert_to_newsletter` (and removed the environment check in `upsert`). We get there just fine and toggle the subscription status in the payload:

```ruby
 return unless needs_subscription_update?
```

And a corresponding method roughly like this:

```ruby
def needs_subscription_update?
  saved_changes.key?(:user_email) || saved_changes.key?(:email_newsletter)
end
```

But that made me think: we trigger `MailChimpBot` through a worker, which means by the time we run it, the user has been saved to the DB and loaded again inside the job, at which point the `saved_changes` we assign in the initializer would always be empty:

```ruby
@saved_changes = user.saved_changes
```

So it's probably better to move this into the callback itself:

```ruby
  def subscribe_to_mailchimp_newsletter
    return unless email.present? && email.include?("@")
    return if saved_changes.key?(:unconfirmed_email) && saved_changes.key?(:confirmation_sent_at)
    return if saved_changes.key?(:user_email) || saved_changes.key?(:email_newsletter)

    Users::SubscribeToMailchimpNewsletterWorker.perform_async(id)
  end
```

I'll do the latter now and add a couple of tests to check that we only call `MailChimpBot` when the user's email or subscription status changes.

@maestromac Can you please sanity check everything I wrote above and tell me if I'm missing something? This was written pretty much stream of consciousness, taking notes as I was debugging.

citizen428 commented on spec/factories/users.rb:
```diff
@@ -69,7 +69,7 @@
     end
 
-    trait :ignore_after_callback do
+    trait :ignore_mailchimp_subscribe_callback do
```
@maestromac I thought it's better to give this trait a name that better reflects what it actually does.

maestromac commented on app/models/user.rb:
```diff
@@ -417,6 +417,7 @@ def non_banished_username
   def subscribe_to_mailchimp_newsletter
     return unless email.present? && email.include?("@")
     return if saved_changes["unconfirmed_email"] && saved_changes["confirmation_sent_at"]
+    return if saved_changes.key?(:user_email) || saved_changes.key?(:email_newsletter)
```
Correct me if I'm wrong, so this would exit early if saved_changes include `:email_newsletter`. That would be a problem if a user decides to unsubscribe on our platform as our platform won't update MailChimp.

citizen428 replied:
That should be an `unless` 🤦‍♂

citizen428 commented:
@maestromac All updated now, please review again.
